components/ntp_background_images/browser/ntp_sponsored_rich_media_source.cc

```diff
@@ -0,0 +1,168 @@
+                     weak_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void NTPSponsoredRichMediaSource::ReadFileCallback(
+    GotDataCallback callback,
+    std::optional<std::string> input) {
+  if (!input) {
+    return std::move(callback).Run(scoped_refptr<base::RefCountedMemory>());
+  }
+
+  std::move(callback).Run(
+      new base::RefCountedBytes(base::as_byte_span(*input)));
+}
+
+std::string NTPSponsoredRichMediaSource::GetMimeType(const GURL& url) {
+  std::string mime_type;
+  const base::FilePath::StringType file_path_extension =
+      base::FilePath::FromUTF8Unsafe(url.path_piece()).Extension();
+  if (!file_path_extension.empty()) {
+    net::GetWellKnownMimeTypeFromExtension(file_path_extension.substr(1),
+                                           &mime_type);
+  }
+
+  return mime_type;
+}
+
+bool NTPSponsoredRichMediaSource::AllowCaching() {
+  return false;
+}
+
+std::string NTPSponsoredRichMediaSource::GetContentSecurityPolicy(
+    network::mojom::CSPDirectiveName directive) {
+  switch (directive) {
+    case network::mojom::CSPDirectiveName::FrameAncestors:
+      return std::string("frame-ancestors ") + kBraveUINewTabURL + ";";
+    case network::mojom::CSPDirectiveName::Sandbox:
+      return "sandbox allow-scripts;";
+    case network::mojom::CSPDirectiveName::DefaultSrc:
+      return "default-src 'none';";
+    case network::mojom::CSPDirectiveName::BaseURI:
+      return "base-uri 'none';";
+    case network::mojom::CSPDirectiveName::FormAction:
+      return "form-action 'none';";
+    case network::mojom::CSPDirectiveName::ScriptSrc:
+      return "script-src 'self';";
+    case network::mojom::CSPDirectiveName::StyleSrc:
+      return "style-src 'self';";
+    case network::mojom::CSPDirectiveName::ImgSrc:
+      return "img-src 'self';";
+    case network::mojom::CSPDirectiveName::MediaSrc:
+      return "media-src 'self';";
+    case network::mojom::CSPDirectiveName::RequireTrustedTypesFor:
+      return "require-trusted-types-for 'script';";
+    case network::mojom::CSPDirectiveName::TrustedTypes:
+      return "trusted-types;";
+    default:
+      // Return empty CSP to avoid inheriting potentially permissive defaults
+      // from content::URLDataSource::GetContentSecurityPolicy().
+      return std::string();
+  }
+}
+
+std::optional<base::FilePath>
+NTPSponsoredRichMediaSource::MaybeGetFilePathForRequestPath(
+    const std::string& request_path) {
+  const NTPSponsoredImagesData* const data =
+      service_->GetBrandedImagesData(false);
+  if (!data) {
+    // No data, deny access.
+    return std::nullopt;
+  }
+
+  const base::FilePath request_path_dir_name =
+      base::FilePath::FromUTF8Unsafe(request_path).DirName();
+
+  for (const auto& campaign : data->campaigns) {
+    for (const auto& creative : campaign.backgrounds) {
+      // Sandbox the request to the creative directory to prevent path
+      // traversal, denying access to other directories.
```

Comment:
can you add some test cases for this? not sure if it's complete enough

Reply:
Sure we are merging tests tomorrow which include these. Thanks

Reply:
can we check if request_path references `..` using https://source.chromium.org/chromium/chromium/src/+/main:base/files/file_path.h;l=398;bpv=0;bpt=1 and throw an error if so?

Reply:
Sure will. And merging tests tomorrow. Thanks

Reply:
@tmancey see example from playlist here: https://github.com/brave/brave-core/blob/master/components/playlist/browser/playlist_service.cc#L1217-L1219

Reply:
Fixed